libs/network/network.cpp

```diff
@@ -0,0 +1,95 @@
+#include "pxt.h"
+#include "JDProtocol.h"
+
+namespace network {
+
+class JDProxyDriver : public JDDriver {
+  public:
+    RefCollection *methods;
+
+    JDProxyDriver(JDDevice d, RefCollection *m, int id) : JDDriver(d, id) {
+        this->methods = m;
+        incrRC(m);
+    }
+
+    virtual int handleControlPacket(JDPkt *p) {
+        auto buf = pxt::mkBuffer((const uint8_t*)&p->crc, p->size + 4);
+        auto r = pxt::runAction1(methods->getAt(0), (TValue)buf);
+        auto retVal = numops::toBool(r) ? DEVICE_OK : DEVICE_CANCELLED;
+        decr(r);
+        decrRC(buf);
+        return retVal;
+    }
+
+    virtual int handlePacket(JDPkt *p) {
+        auto buf = pxt::mkBuffer((const uint8_t*)&p->crc, p->size + 4);
+        auto r = pxt::runAction1(methods->getAt(1), (TValue)buf);
+        auto retVal = numops::toBool(r) ? DEVICE_OK : DEVICE_CANCELLED;
+        decr(r);
+        decrRC(buf);
+        return retVal;
+    }
+
+    virtual int fillControlPacket(JDPkt *p) {
+        auto buf = pxt::mkBuffer((const uint8_t*)&p->crc, JD_SERIAL_DATA_SIZE + 4);
+        auto r = pxt::runAction1(methods->getAt(2), (TValue)buf);
+        memcpy(&p->crc, buf->data, JD_SERIAL_DATA_SIZE + 4);
+        (void)r;
+        // decr(r); // TODO compiler might return non-null on void functions, so better skip decr
+        decrRC(buf);
+        return DEVICE_OK;
+    }
+
+    virtual int deviceConnected(JDDevice device) {
+        auto r = JDDriver::deviceConnected(device);
+        pxt::runAction0(methods->getAt(3));
+        return r;
+    }
+
+    virtual int deviceRemoved() {
+        auto r = JDDriver::deviceRemoved();
+        pxt::runAction0(methods->getAt(4));
+        return r;
+    }
+
+    JDDevice *getDevice() { return &device; }
+
+    ~JDProxyDriver() { decrRC(methods); }
+};
+
+//%
+JDProxyDriver *addNetworkDriver(int driverClass, RefCollection *methods) {
+    static int deviceId = 3030;
+    return new JDProxyDriver(JDDevice(0, 0, 0, driverClass), methods, ++deviceId);
```

Comment:
You'll need to pass flags which tells the logic driver what mode to run the driver in. I've got a commit that has an enum instead of a #define for the flags.

Reply:
Sure thing. I'm assuming you're handling the serial number?

Reply:
Will do.